bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java

```diff
@@ -571,45 +568,37 @@ private void publishSuspectedLedgers(String bookieIP, Set<Long> ledgers)
         }
 
         public void operationComplete(int rc, Set<LedgerFragment> fragments) {
-            try {
-                if (rc == BKException.Code.OK) {
-                    Set<BookieSocketAddress> bookies = Sets.newHashSet();
-                    for (LedgerFragment f : fragments) {
-                        bookies.addAll(f.getAddresses());
-                    }
-                    for (BookieSocketAddress bookie : bookies) {
-                        publishSuspectedLedgers(bookie.toString(), Sets.newHashSet(lh.getId()));
-                    }
-                }
-                lh.close();
-            } catch (BKException bke) {
-                LOG.error("Error closing lh", bke);
-                if (rc == BKException.Code.OK) {
-                    rc = BKException.Code.ReplicationException;
-                }
-            } catch (InterruptedException ie) {
-                LOG.error("Interrupted publishing suspected ledger", ie);
-                Thread.currentThread().interrupt();
-                if (rc == BKException.Code.OK) {
-                    rc = BKException.Code.InterruptedException;
-                }
-            } catch (BKAuditException bkae) {
-                LOG.error("Auditor exception publishing suspected ledger", bkae);
-                if (rc == BKException.Code.OK) {
-                    rc = BKException.Code.ReplicationException;
+            if (rc == BKException.Code.OK) {
+                Set<BookieSocketAddress> bookies = Sets.newHashSet();
+                for (LedgerFragment f : fragments) {
+                    bookies.addAll(f.getAddresses());
                 }
+                publishSuspectedLedgersAsync(
+                    bookies.stream().map(BookieSocketAddress::toString).collect(Collectors.toList()),
+                    Sets.newHashSet(lh.getId())
+                ).whenComplete((result, cause) -> {
+                    if (null != cause) {
+                        callback.processResult(Code.ReplicationException, null, null);
+                    } else {
+                        callback.processResult(Code.OK, null, null);
+                    }
+                });
+            } else {
+                callback.processResult(rc, null, null);
             }
-
-            callback.processResult(rc, null, null);
+            lh.closeAsync().whenComplete((result, cause) -> {
```

Comment:
why is it responsibility of ProcessLostFragmentsCb to close the ledgerhandle, I'm not sure if it is appropriate to do it here, since ProcessLostFragmentsCb is not the owner of 'lh'

Reply:
I didn’t change the behavior. The original logic is ProcessLostFramgmentsCb closing it.

Reply:
checkLedgersProcessor 
 closes the 'lh' https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java#L679

Reply:
check here: https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java#L584